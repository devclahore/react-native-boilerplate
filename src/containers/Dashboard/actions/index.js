/*
  Author: Taimoor Khan
  GitHub: https://github.com/Taimoormk
  Email: taimoor.m.k AT LIVE.COM 
*/

// ########## Import Dependencies Here ##########

// ########## Import Components Here ##########
import * as types from "../constants";
import apiCaller from "../../../config/apiCaller";
import ENDPOINTS, { REQUEST_TYPES } from "../../../config/endpoints";

// ########## Import MainScreen Actions Here ##########
export const dashboardSuccessAction = data => ({
  type: types.DASHBOARD_SUCCESS,
  payload: data
});

export const dashboardFailAction = err => ({
  type: types.DASHBOARD_FAIL,
  payload: err
});

// ########## Thunk Logic Goes Here ##########
export const dashboardAttemptAction = username => {
  const thunk = async function thunk(dispatch) {
    try {
      const data = await apiCaller(REQUEST_TYPES.GET, `${ENDPOINTS.GITHUB_USERS}/${username}`, null, null);
      dispatch(dashboardSuccessAction(data));
    } catch (err) {
      dispatch(dashboardFailAction(err));
    }
  };
  return thunk;
};
