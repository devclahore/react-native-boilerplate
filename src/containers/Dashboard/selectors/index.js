/*
  Author: Taimoor Khan
  GitHub: https://github.com/Taimoormk
  Email: taimoor.m.k AT LIVE.COM 
*/

// ########## Import Dependencies Here ##########
import { createSelector } from "reselect";

export const getDashboard = state => state.dashboardReducer.data;

export const getUserName = createSelector(
  [ getDashboard ],
  data => data.name
);

export const getUserBio = createSelector(
  [ getDashboard ],
  data => data.bio
);

export const getUserCompany = createSelector(
  [ getDashboard ],
  data => data.company
);

export const getUserLocation = createSelector(
  [ getDashboard ],
  data => data.location
);

export const getUserAvatarUrl = createSelector(
  [ getDashboard ],
  data => data.avatarUrl
);
