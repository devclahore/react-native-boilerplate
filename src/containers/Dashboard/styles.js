/*
  Author: Taimoor Khan
  GitHub: https://github.com/Taimoormk
  Email: taimoor.m.k AT LIVE.COM 
*/

// ########## Import Dependencies Here ##########
import styled from "styled-components/native";

import { STYLE_CONSTANTS, width } from "../../config/styles";

export const ContainerView = styled.View`
  padding: 10px;
`;

export const FlexContainerView = styled.View`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const H1Text = styled.Text`
  font-family: ${STYLE_CONSTANTS.H1.fontFamily};
  font-size: ${STYLE_CONSTANTS.H1.fontSize};
  color: ${STYLE_CONSTANTS.H1.color};
  text-transform: ${STYLE_CONSTANTS.H1.textTransform};
  margin-top: 10;
  margin-bottom: 10;
`;
