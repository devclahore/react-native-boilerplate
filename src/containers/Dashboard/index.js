/*
  Author: Taimoor Khan
  GitHub: https://github.com/Taimoormk
  Email: taimoor.m.k AT LIVE.COM 
*/

// ########## Import Dependencies Here ##########
import React, { Component, Fragment } from "react";
import { string, func, oneOfType, any } from "prop-types";
import { connect } from "react-redux";
import { View, Text, Image } from "react-native";
import { TextInput, Button, Headline, Title } from "react-native-paper";

// ########## Import Containers Here ##########

// ########## Import Components Here ##########
import styles from "./styles";
import * as actions from "./actions";
import { STYLE_CONSTANTS } from "../../config/styles";
import * as selectors from "./selectors";

class Dashboard extends Component {
  state = {
    textInput: ""
  };

  onPressHandler = () => {
    const { textInput } = this.state;
    const { dashboardAttempt } = this.props;
    dashboardAttempt(textInput);
  };

  render() {
    const { textInput } = this.state;
    const { name, bio, company, location, avatarUrl } = this.props;
    return (
      <Fragment>
        <View style={styles.container}>
          <View style={styles.textWrapper}>
            <Text style={styles.heading}>Dashboard</Text>
          </View>
          <View style={styles.form}>
            <TextInput
              mode="outlined"
              label="Enter GitHub handle"
              autoCapitalize="none"
              value={textInput}
              style={styles.input}
              onChangeText={val => this.setState({ textInput: val })}
            />
            <Button
              icon="add-a-photo"
              mode="contained"
              color={STYLE_CONSTANTS.COLORS.SECONDARY}
              onPress={this.onPressHandler}
            >
              Fetch Repos
            </Button>
            {name && (
              <View style={styles.userDataContainer}>
                <View style={styles.avatarContainer}>
                  <Image
                    style={styles.userAvatar}
                    source={{ uri: avatarUrl }}
                  />
                </View>
                <View style={styles.userContainer}>
                  <Headline style={styles.userData}>{name}</Headline>
                  <Title style={styles.userData}>{bio}</Title>
                  <Title style={styles.userData}>{company}</Title>
                  <Title style={styles.userData}>{location}</Title>
                </View>
              </View>
            )}
          </View>
        </View>
      </Fragment>
    );
  }
}

Dashboard.propTypes = {
  name: oneOfType([
    string.isRequired,
    any,
  ]),
  bio: oneOfType([
    string.isRequired,
    any,
  ]),
  company: oneOfType([
    string.isRequired,
    any,
  ]),
  location: oneOfType([
    string.isRequired,
    any,
  ]),
  avatarUrl: oneOfType([
    string.isRequired,
    any,
  ]),
  dashboardAttempt: func.isRequired
};

const mapStateToProps = state => {
  if (state.dashboardReducer.data) {
    return {
      name: selectors.getUserName(state),
      bio: selectors.getUserBio(state),
      company: selectors.getUserCompany(state),
      location: selectors.getUserLocation(state),
      avatarUrl: selectors.getUserAvatarUrl(state),
    }
  } else {
    return {
      name: null,
      bio: null,
      company: null,
      location: null,
      avatarUrl: null,
    };
  }
};

const mapDispatchToProps = dispatch => {
  return {
    dashboardAttempt: username =>
      dispatch(actions.dashboardAttemptAction(username))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Dashboard);
