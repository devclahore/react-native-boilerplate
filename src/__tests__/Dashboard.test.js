/* eslint-disable no-undef */
/*
  Author: Taimoor Khan
  GitHub: https://github.com/Taimoormk
  Email: taimoor.m.k AT LIVE.COM 
*/

// ########## Import Dependencies Here ##########
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import configureStore from 'redux-mock-store';

// ########## Import Components Here ##########
import Dashboard from '../containers/Dashboard';

// ########## Create Initial State ##########
const fakeInitialState = {
  dashboardReducer: {
    data: null,
    error: null,
  },
};

const mockStore = configureStore();

test('renders Dashboard componenent correctly', () => {
  const renderer = new ShallowRenderer();
  const store = mockStore(fakeInitialState);
  const dashboardAttempt = () => null;
  renderer.render(<Dashboard store={store} dashboardAttempt={dashboardAttempt} />);
  const wrapper = renderer.getRenderOutput();
  expect(wrapper).toMatchSnapshot();
});
