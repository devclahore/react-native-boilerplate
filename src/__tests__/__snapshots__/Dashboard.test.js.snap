// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Dashboard componenent correctly 1`] = `
<Context.Provider
  value={null}
>
  <Dashboard
    dashboardAttempt={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    userdata={null}
  />
</Context.Provider>
`;
